PyDynamic/uncertainty/propagate_DFT.py

```diff
@@ -1,776 +1,908 @@
+
+    If `return_type` is `separate`:
+
+    Returns
+    -------
+        A: np.ndarray
+            vector of magnitude values
+        P: np.ndarray
+            vector of phase values in radians, in the range [-pi, pi]
+        UAP: np.ndarray
+            covariance matrix associated with (A,P)
+
+    Otherwise:
+
+    Returns
+    -------
+        AP: np.ndarray
+            vector of magnitude and phase values
+        UAP: np.ndarray
+            covariance matrix associated with AP
+    """
+    # calculate inverse DFT
+    N = len(F) - 2
+    R = F[:N // 2 + 1]
+    I = F[N // 2 + 1:]
+
+    A = np.sqrt(R ** 2 + I ** 2)  # absolute value
+    P = np.arctan2(I, R)  # phase value
+    if len(UF.shape) == 1:
+        uF = 0.5 * (np.sqrt(UF[:N // 2 + 1]) + np.sqrt(
+            UF[N // 2 + 1:]))  # uncertainty of real,imag
+    else:
+        uF = 0.5 * (np.sqrt(np.diag(UF[:N // 2 + 1, :N // 2 + 1])) + np.sqrt(
+            np.diag(UF[N // 2 + 1:, N // 2 + 1:])))
+    if np.any(A / uF < tol):
+        print(
+            'DFT2AmpPhase Warning\n Some amplitude values are below the '
+            'defined threshold.')
+        print(
+            'The GUM formulas may become unreliable and a Monte Carlo '
+            'approach is recommended instead.')
+        print(
+            'The actual minimum value of A/uF is %.2e and the threshold is '
+            '%.2e' % (
+                (A / uF).min(), tol))
+    aR = R / A  # sensitivities
+    aI = I / A
+    pR = -I / A ** 2
+    pI = R / A ** 2
+
+    if len(UF.shape) == 1:  # uncertainty calculation of zero correlation
+        URR = UF[:N // 2 + 1]
+        UII = UF[N // 2 + 1:]
+        U11 = URR * aR ** 2 + UII * aI ** 2
+        U12 = aR * URR * pR + aI * UII * pI
+        U22 = URR * pR ** 2 + UII * pI ** 2
+        UAP = sparse.diags([np.r_[U11, U22], U12, U12],
+                           [0, N // 2 + 1, -(N // 2 + 1)])
+        if not keep_sparse:
+            UAP = UAP.toarray()
+    else:  # uncertainty calculation for full covariance
+        URR = UF[:N // 2 + 1, :N // 2 + 1]
+        URI = UF[:N // 2 + 1, N // 2 + 1:]
+        UII = UF[N // 2 + 1:, N // 2 + 1:]
+        U11 = _prod(aR, _prod(URR, aR)) + _prod(aR, _prod(URI, aI)) +\
+            _prod(aI, _prod(URI.T, aR)) + _prod(aI, _prod(UII, aI))
+        U12 = _prod(aR, _prod(URR, pR)) + _prod(aI, _prod(URI, pI)) +\
+            _prod(aI, _prod(URI.T, pR)) + _prod(aI, _prod(UII, pI))
+        U22 = _prod(pR, _prod(URR, pR)) + _prod(pR, _prod(URI, pI)) +\
+            _prod(pI, _prod(URI.T, pR)) + _prod(pI, _prod(UII, pI))
+        UAP = np.vstack((np.hstack((U11, U12)), np.hstack((U12.T, U22))))
+
+    if return_type == "separate":
+        return A, P, UAP  # amplitude and phase as separate variables
+    else:
+        return np.r_[A, P], UAP
+
+
+def AmpPhase2DFT(A, P, UAP, keep_sparse=False):
+    """Transformation from magnitude and phase to real and imaginary parts
```

Comment:
In many of the other functions, >>"""<< was followed by a newline. Maybe here to?

Reply:
All oneliners should stay on the same line as the triple quote, so I changed the rest of the file.